Kin Android/app/src/main/java/kin/unity/BackupActivity.java

```diff
@@ -0,0 +1,179 @@
+
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        Intent intent = getIntent();
+        String action = intent.getAction();
+        clientId = intent.getStringExtra("clientId");
+
+        if (action.equals(KinPlugin.BACKUP_ACTION))
+        {
+            accountId = intent.getStringExtra("accountId");
+            backupAccount(accountId, clientId);
+        }
+        else if (action.equals(KinPlugin.RESTORE_ACTION))
+        {
+            restoreAccount(clientId);
+        }
+    }
+
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        super.onActivityResult(requestCode, resultCode, data);
+        if (requestCode == REQ_CODE_BACKUP || requestCode == REQ_CODE_RESTORE) {
+            backupManager.onActivityResult(requestCode, resultCode, data);
+            finish();
+        }
+    }
+
+    @Override
+    protected void onDestroy() {
+        super.onDestroy();
+        if (backupManager != null)
+        {
+            releaseBackupManager();
+        }
+    }
+
+    public BackupAndRestoreManager createBackupAndRestoreManager(String clientId, String accountId)
+    {
+        BackupAndRestoreManager backupManager = new BackupAndRestoreManager
+                (this, REQ_CODE_BACKUP, REQ_CODE_RESTORE);
+
+        backupManager.registerBackupCallback(new BackupCallback() {
+            @Override
+            public void onSuccess() {
+                Log.d(LOGTAG, "Got success from backup manager");
+                _plugin.unitySendMessage("BackupSucceeded", _plugin.callbackToJson(null, accountId));
+            }
+
+            @Override
+            public void onCancel() {
+                Log.d(LOGTAG, "Got cancel from backup manager");
+                _plugin.unitySendMessage("BackupCanceled", _plugin.callbackToJson(null, accountId));
+            }
+
+            @Override
+            public void onFailure(BackupAndRestoreException e) {
+                Log.e(LOGTAG, "Got failure from backup manager");
+                _plugin.unitySendMessage("BackupFailed", _plugin.exceptionToJson(e, accountId));
+            }
+        });
+
+        backupManager.registerRestoreCallback(new RestoreCallback() {
+            @Override
+            public void onSuccess(KinClient kinClient, KinAccount kinAccount) {
+                Log.d(LOGTAG, "Got success from backup manager");
+                String accountId = _plugin.getAccountIdByAccount(kinAccount);
+                // If we already have this account, there is no need to add it again to the list.
+                // The account will still be added internally to the kin client until this bug will be fixed in the android b&r module
+                if (accountId == null)
+                {
+                    accountId = _plugin.generateUniqueId();
+                    _plugin._accounts.put(accountId, kinAccount);
```

Comment:
why not use the public key as unique id/key?

Reply:
Usually, the id is generated in c#, before you have the address, so Mike chose to use some arbitrary id